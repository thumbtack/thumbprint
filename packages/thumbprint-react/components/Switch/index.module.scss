@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';

// Match the width/height of Thubmprint Checkbox
$bar-height: 20px;
$bar-height-large: 28px;
$height-width-ratio: 1.8;
$transition-duration: 0.1s;
$transition-function: ease-in;

.slideToggle {
    width: $height-width-ratio * $bar-height;
    height: $bar-height;
    border-radius: $bar-height;
    cursor: pointer;
    transition: background-color $transition-duration $transition-function;

    &Large {
        width: $height-width-ratio * $bar-height-large;
        height: $bar-height-large;
    }

    &Disabled {
        user-select: none;
        cursor: default;
        opacity: 0.3;
    }
}

.toggleHandle {
    height: $bar-height - 2px; // Subtract 2px to account for top/bottom border of parent container
    width: $bar-height - 2px; // Subtract 2px to account for left/right border of parent container
    border-radius: 50%;
    transition: transform $transition-duration $transition-function;
    box-shadow: $tp-shadow__card;
    box-sizing: border-box;

    &Large {
        height: $bar-height-large - 2px; // Subtract 2px to account for top/bottom border of parent container
        width: $bar-height-large - 2px; // Subtract 2px to account for left/right border of parent container
    }

    .slideToggleChecked & {
        transform: translate3d(($height-width-ratio - 1) * $bar-height, 0, 0);
    }

    .slideToggleLargeChecked & {
        transform: translate3d(($height-width-ratio - 1) * $bar-height-large, 0, 0);
    }

    input:focus ~ & {
        box-shadow: 0 0 1px 3px $tp-color__gray-300;
    }
}
