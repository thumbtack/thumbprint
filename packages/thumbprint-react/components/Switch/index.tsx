import React from 'react';
import classNames from 'classnames';
import styles from './index.module.scss';

export const SWITCH_COLOR = {
    blue: 'blue',
    green: 'green',
};

export const SWITCH_SIZES = {
    small: 'small',
    large: 'large',
};

export interface SwitchProps {
    /**
     * The `id` is added to the switch as an HTML attribute and passed to the `onChange`
     * function.
     */
    id?: string;
    /**
     * Determines if the switch is checked.
     */
    isChecked?: boolean;
    /**
     * Function that runs when a switch value changes. It receives the new boolean value,
     * and the provided `id` as such:
     * `props.onChange(event.target.checked, props.id)`.
     */
    onChange: (isChecked: boolean, id: string) => void;
    /**
     * Disables the component.
     */
    isDisabled?: boolean;
    /**
     * Determines how the switch sizes. The default size is `small`.
     */
    size?: keyof typeof SWITCH_SIZES;
    /**
     * This adds an `aria-label` to the element.
     * the importance of using labels](https://dequeuniversity.com/rules/axe/4.3/select-name).
     */
    accessibilityLabel?: string;
    /**
     * Determines how the switch colors. The default color is `blue`.
     */
    color?: keyof typeof SWITCH_COLOR;
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTestId?: string;
    /**
     * Switches on a page with the same name will be grouped together when sent to the server.
     */
    name?: string;
}

export default function Switch({
    id = '',
    isDisabled = false,
    isChecked = false,
    onChange,
    dataTestId,
    name,
    accessibilityLabel,
    color = SWITCH_COLOR.blue as keyof typeof SWITCH_COLOR,
    size = SWITCH_SIZES.small as keyof typeof SWITCH_SIZES,
}: SwitchProps): JSX.Element {
    const bgColor = `bg-${color} b-${color}`;

    const onToggle = (): void => {
        onChange(!isChecked, id);
    };
    return (
        <div
            className={classNames(styles.slideToggle, 'ba', {
                [styles.slideToggleLarge]: size === SWITCH_SIZES.large,
                [styles.slideToggleDisabled]: isDisabled,
                [styles.slideToggleChecked]: isChecked,
                [styles.slideToggleLargeChecked]: size === SWITCH_SIZES.large && isChecked,
                [bgColor]: isChecked,
                'bg-gray b-gray-300': !isChecked,
            })}
            onClick={isDisabled ? undefined : onToggle}
            onKeyUp={
                isDisabled
                    ? undefined
                    : (event): void => {
                          if (event.key === 'Enter' || event.keyCode === 13) {
                              onChange(!isChecked, id);
                          }
                      }
            }
        >
            <input
                className="visually-hidden"
                type="checkbox"
                role="switch"
                id={id}
                data-testid={dataTestId}
                disabled={isDisabled}
                checked={isChecked}
                name={name}
                aria-checked={isChecked}
                aria-disabled={isDisabled}
                aria-label={accessibilityLabel}
                onChange={isDisabled ? undefined : onToggle}
            />

            <div
                className={classNames(styles.toggleHandle, 'bg-white ba b-gray-300', {
                    [styles.toggleHandleLarge]: size === SWITCH_SIZES.large,
                })}
            />
        </div>
    );
}
