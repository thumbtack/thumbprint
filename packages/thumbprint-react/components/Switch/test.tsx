import React from 'react';
import { mount } from 'enzyme';
import Switch from './index';

test('renders Switch input', () => {
    const wrapper = mount(<Switch onChange={jest.fn()} />);
    expect(wrapper.find('input').prop('type')).toBe('checkbox');
    expect(wrapper).toMatchSnapshot();
});

test('adds `disabled` attribute', () => {
    const wrapper = mount(<Switch isDisabled onChange={jest.fn()} />);
    expect(wrapper.find('input').prop('disabled')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('adds `name` attribute', () => {
    const wrapper = mount(<Switch name="ohhi" onChange={jest.fn()} />);
    expect(wrapper.find('input').prop('name')).toBe('ohhi');
    expect(wrapper).toMatchSnapshot();
});

test('is not checked by default', () => {
    const wrapper = mount(<Switch onChange={jest.fn()} />);
    expect(wrapper.find('input').prop('checked')).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
});

test('is checked when `isChecked` is set to true', () => {
    const wrapper = mount(<Switch isChecked onChange={jest.fn()} />);
    expect(wrapper.find('input').prop('checked')).toBe(true);
    expect(wrapper).toMatchSnapshot();
});

test('calls `onChange` function when supplied', () => {
    const onChange = jest.fn();

    const wrapper = mount(<Switch onChange={onChange} />);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledTimes(1);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledTimes(2);
});

test('adds `dataTestId` prop', () => {
    const wrapper = mount(<Switch dataTestId="Duck" onChange={jest.fn()} />);
    expect(wrapper.find('input').prop('data-testid')).toEqual('Duck');
    expect(wrapper).toMatchSnapshot();
});

test('`value` attribute is `"on"` in form data if prop is `undefined` and Switch is checked', () => {
    let nameInFormData;
    const name = 'foo';

    const wrapper = mount(
        <form
            onSubmit={(e): void => {
                e.preventDefault();
                // We need to assert the type because TypeScript can't infer it.
                // https://stackoverflow.com/a/43851475
                const data = new FormData(e.target as HTMLFormElement);
                nameInFormData = data.get(name);
            }}
        >
            <Switch isChecked name={name} onChange={(): void => {}} />
            <button type="submit">Submit</button>
        </form>,
    );

    wrapper.find('form').simulate('submit');
    expect(nameInFormData).toBe('on');
});

test('`value` attribute is `"on"` in form data if Switch is checked', () => {
    let nameInFormData;
    const name = 'foo';

    const wrapper = mount(
        <form
            onSubmit={(e): void => {
                e.preventDefault();
                // We need to assert the type because TypeScript can't infer it.
                // https://stackoverflow.com/a/43851475
                const data = new FormData(e.target as HTMLFormElement);
                nameInFormData = data.get(name);
            }}
        >
            <Switch isChecked name={name} onChange={(): void => {}} />
            <button type="submit">Submit</button>
        </form>,
    );

    wrapper.find('form').simulate('submit');
    expect(nameInFormData).toBe('on');
});
