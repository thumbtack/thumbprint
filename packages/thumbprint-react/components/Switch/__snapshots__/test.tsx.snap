// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<Wrap
  dataTest="Duck"
>
  <div
    className="root"
    data-test="Duck"
  >
    goose
  </div>
</Wrap>
`;

exports[`adds \`dataTestId\` prop 1`] = `
<Switch
  dataTestId="Duck"
  onChange={[MockFunction]}
>
  <div
    className="slideToggle ba bg-gray b-gray-300"
    onClick={[Function]}
    onKeyUp={[Function]}
  >
    <input
      aria-checked={false}
      aria-disabled={false}
      checked={false}
      className="visually-hidden"
      data-testid="Duck"
      disabled={false}
      id=""
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="toggleHandle bg-white ba b-gray-300"
    />
  </div>
</Switch>
`;

exports[`adds \`disabled\` attribute 1`] = `
<Switch
  isDisabled={true}
  onChange={[MockFunction]}
>
  <div
    className="slideToggle ba slideToggleDisabled bg-gray b-gray-300"
  >
    <input
      aria-checked={false}
      aria-disabled={true}
      checked={false}
      className="visually-hidden"
      disabled={true}
      id=""
      role="switch"
      type="checkbox"
    />
    <div
      className="toggleHandle bg-white ba b-gray-300"
    />
  </div>
</Switch>
`;

exports[`adds \`name\` attribute 1`] = `
<Switch
  name="ohhi"
  onChange={[MockFunction]}
>
  <div
    className="slideToggle ba bg-gray b-gray-300"
    onClick={[Function]}
    onKeyUp={[Function]}
  >
    <input
      aria-checked={false}
      aria-disabled={false}
      checked={false}
      className="visually-hidden"
      disabled={false}
      id=""
      name="ohhi"
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="toggleHandle bg-white ba b-gray-300"
    />
  </div>
</Switch>
`;

exports[`is checked when \`isChecked\` is set to true 1`] = `
<Switch
  isChecked={true}
  onChange={[MockFunction]}
>
  <div
    className="slideToggle ba slideToggleChecked bg-blue b-blue"
    onClick={[Function]}
    onKeyUp={[Function]}
  >
    <input
      aria-checked={true}
      aria-disabled={false}
      checked={true}
      className="visually-hidden"
      disabled={false}
      id=""
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="toggleHandle bg-white ba b-gray-300"
    />
  </div>
</Switch>
`;

exports[`is not checked by default 1`] = `
<Switch
  onChange={[MockFunction]}
>
  <div
    className="slideToggle ba bg-gray b-gray-300"
    onClick={[Function]}
    onKeyUp={[Function]}
  >
    <input
      aria-checked={false}
      aria-disabled={false}
      checked={false}
      className="visually-hidden"
      disabled={false}
      id=""
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="toggleHandle bg-white ba b-gray-300"
    />
  </div>
</Switch>
`;

exports[`renders Switch input 1`] = `
<Switch
  onChange={[MockFunction]}
>
  <div
    className="slideToggle ba bg-gray b-gray-300"
    onClick={[Function]}
    onKeyUp={[Function]}
  >
    <input
      aria-checked={false}
      aria-disabled={false}
      checked={false}
      className="visually-hidden"
      disabled={false}
      id=""
      onChange={[Function]}
      role="switch"
      type="checkbox"
    />
    <div
      className="toggleHandle bg-white ba b-gray-300"
    />
  </div>
</Switch>
`;

exports[`renders content that is passed in 1`] = `
<Wrap>
  <div
    className="root"
  >
    goose
  </div>
</Wrap>
`;

exports[`renders content that is passed in with a bleed breakpoint 1`] = `
<Wrap
  bleedBelow="medium"
>
  <div
    className="root bleedBelowMedium"
  >
    goose
  </div>
</Wrap>
`;
